### PR TITLE
LM Optimizer

### PR DESCRIPTION
**Pull request by jumpinthefire**
Initial implementation of LMA. Reference: https://github.com/ceres-solver/ceres-solver/blob/master/include/ceres/tiny_solver.h

Tested with `m3500_benchmark` and `small_problem`.

### CODE REVIEW

File: src/optimizer/levenberg_marquardt_optimizer.rs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+        // On the first iteration, we'll generate a diagonal matrix of the jacobian.
+        // Its shape will be (total_variable_dimension, total_variable_dimension).
+        // With LM, rather than solving A * dx = b for dx, we solve for (A + lambda * diag(A)) dx = b.
+        let mut jacobi_scaling_diagonal: Option<faer::sparse::SparseColMat<usize, f64>> = None;
+
+        const MIN_DIAGONAL: f64 = 1e-6;
+        const MAX_DIAGONAL: f64 = 1e32;
+
+        const INITIAL_TRUST_REGION_RADIUS: f64 = 1e4;
```
Comment (powei-lin) [inline src/optimizer/levenberg_marquardt_optimizer.rs]: Can you move these to `LevenbergMarquardtOptimizer`. And `impl Default for LevenbergMarquardtOptimizer` to use these default values?

File: src/optimizer/levenberg_marquardt_optimizer.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+                .mul(jac.as_ref());
+
+            // J^T * -r = Matrix of shape (total_variable_dimension, 1)
+            let jtr = jac.as_ref().transpose().mul(-&residuals);
+
+            // Regularize the diagonal of jtj between MIN_DIAGONAL and MAX_DIAGONAL.
+            let mut jtj_regularized = jtj.clone();
+            for i in 0..problem.total_variable_dimension {
+                jtj_regularized[(i, i)] =
+                    (jtj.get(i, i).unwrap().max(MIN_DIAGONAL)).min(MAX_DIAGONAL);
```
Comment (powei-lin) [inline src/optimizer/levenberg_marquardt_optimizer.rs]: The shape of the matrix is always the same. Is there an unsafe get mut to accelerate this part?

File: src/optimizer/levenberg_marquardt_optimizer.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+                // Compute residuals of (x + dx)
+                let new_residuals = problem.compute_residuals(&new_params, true);
+
+                // rho is the ratio between the actual reduction in error and the reduction
+                // in error if the problem were linear.
+                let actual_residual_change =
+                    &residuals.squared_norm_l2() - &new_residuals.squared_norm_l2();
+                let linear_residual_change: faer::Mat<f64> =
+                    lm_step.adjoint().mul(2.0 * &jtr - &jtj * &lm_step);
```
Comment (powei-lin) [inline src/optimizer/levenberg_marquardt_optimizer.rs]: Why are you using adjoint here?

**Comment by powei-lin** [comment]
Can you bump the minor version, please?

File: src/optimizer/levenberg_marquardt_optimizer.rs (hunk 1 above)
Comment (jumpinthefire) [reply src/optimizer/levenberg_marquardt_optimizer.rs]: ✅

File: src/optimizer/levenberg_marquardt_optimizer.rs (hunk 2 above)
Comment (jumpinthefire) [reply src/optimizer/levenberg_marquardt_optimizer.rs]: Changed it to `[(i, i)]`, which just panics if the access is out of bounds.

File: src/optimizer/levenberg_marquardt_optimizer.rs (hunk 3 above)
Comment (jumpinthefire) [reply src/optimizer/levenberg_marquardt_optimizer.rs]: Whoops, should be `transpose()`.

**Pull request merged**